bql/semantic/semantic_test.go

```diff
@@ -259,6 +259,31 @@ func TestGraphClauseManipulation(t *testing.T) {
 	}
 }
 
+func TestFilterClauseManipulation(t *testing.T) {
+	st := &Statement{}
+	if wf := st.WorkingFilter(); wf != nil {
+		t.Fatalf(`semantic.Statement.WorkingFilter() = %q for statement "%v" without initialization; want nil`, wf, st)
+	}
+	st.ResetWorkingFilterClause()
+	if wf := st.WorkingFilter(); wf == nil || !wf.IsEmpty() {
+		t.Fatalf(`semantic.Statement.WorkingFilter() = %q for statement "%v" after call to ResetWorkingFilterClause; want empty FilterClause`, wf, st)
+	}
+	st.AddWorkingFilterClause()
+	if got, want := len(st.FilterClauses()), 0; got != want {
+		t.Fatalf(`len(semantic.Statement.FilterClauses()) = %d for statement "%v"; want %d`, got, st, want)
+	}
+
+	wf := st.WorkingFilter()
+	*wf = FilterClause{Operation: "latest", Binding: "?p"}
+	st.AddWorkingFilterClause()
+	if got, want := len(st.FilterClauses()), 1; got != want {
+		t.Fatalf(`len(semantic.Statement.FilterClauses()) = %d for statement "%v"; want %d`, got, st, want)
+	}
+	if wf = st.WorkingFilter(); wf == nil || !wf.IsEmpty() {
+		t.Fatalf(`semantic.Statement.WorkingFilter() = %q for statement "%v"; want empty FilterClause`, wf, st)
+	}
+}
```

Comment:
it seems to me that, although the expectations are clear in the test, the reasoniung for the expectations are not.
To improve it I think you could wrap things in t.Run and have a name for each bit that would explain what you are actually testing. FilterClauseManipulation is too generic.
and, why do you want len == 0 after the first AddWorkingFilterClause, but 1 after the second is not immediately clear.
Although this could be due to the naming of AddWorkingFilterClause which might not be very intuitive.

Reply:
Here I was following the pattern used in a very similar test function `TestGraphClauseManipulation` in this same file. But I do like the idea of using `t.Run` to make the steps being tested clearer here, and I have just done so, thank you Kloss. =]

About your second comment, I agree with you. Inspired by the `AddWorkingGraphClause` method, already present in BadWolf, I had added a check for `nil` and empty `workingFilter` in the `AddWorkingFilterClause` method as well (to avoid an invalid addition).

But these checks made sense only for `AddWorkingGraphClause` since it is activated by the correspondent hook for multiple symbols (as specified by the end of `grammar.go`). Since `AddWorkingFilterClause` is uniquely activated by the hook for the `"FILTER_CLAUSE"` symbol at the moment it finds a closing parenthesis `)`, and the checks for `nil` and empty `workingFilter` are already made in `hooks.go` (inside `whereFilterClause`), there was no need to overcomplicate it with those additional checks. Then, I have just decided to remove them.